File: clients/java/src/main/java/org/aigen/oabp/OabpException.java
package org.aigen.oabp;

/**
 * Checked exception for every failure surfaced by {@link OabpClient}.
 *
 * <p>One checked type covers the three failure families a caller must handle:
 * <ul>
 *   <li><b>transport</b> — the request never produced an HTTP response (I/O error,
 *       connection refused, timeout, interruption);</li>
 *   <li><b>protocol</b> — a response arrived but could not be parsed or was malformed;</li>
 *   <li><b>API</b> — the server returned a non-2xx status (see {@link ApiException},
 *       which exposes the status code and raw body).</li>
 * </ul>
 *
 * <p>It is checked deliberately: talking to a remote service can always fail, and the
 * SDK forces callers to acknowledge that at compile time rather than discover it at
 * runtime.
 */
public class OabpException extends Exception {

    private static final long serialVersionUID = 1L;

    public OabpException(String message) {
        super(message);
    }

    public OabpException(String message, Throwable cause) {
        super(message, cause);
    }

    /**
     * Thrown when the OABP server returns a non-2xx HTTP status.
     *
     * <p>Carries the {@link #statusCode()} and the raw response {@link #body()} so callers
     * can branch on, log, or surface the server's error detail.
     */
    public static class ApiException extends OabpException {

        private static final long serialVersionUID = 1L;

        private final int statusCode;
        private final transient String body;

        public ApiException(int statusCode, String body, String message) {
            super(message);
            this.statusCode = statusCode;
            this.body = body;
        }

        /** @return the HTTP status code returned by the server. */
        public int statusCode() {
            return statusCode;
        }

        /** @return the raw response body (possibly empty, never {@code null}). */
        public String body() {
            return body == null ? "" : body;
        }

        /** @return {@code true} for 4xx statuses (the request was at fault). */
        public boolean isClientError() {
            return statusCode >= 400 && statusCode < 500;
        }

        /** @return {@code true} for 5xx statuses (the server failed). */
        public boolean isServerError() {
            return statusCode >= 500 && statusCode < 600;
        }

        /** @return {@code true} if the resource was not found (HTTP 404). */
        public boolean isNotFound() {
            return statusCode == 404;
        }
    }
}

File: clients/kotlin/src/main/kotlin/org/aigen/oabp/OabpException.kt
package org.aigen.oabp

/**
 * Base type for every failure surfaced by [OabpClient].
 *
 * Three failure families a caller must handle:
 *  - **transport** — the request never produced an HTTP response (I/O error, connection
 *    refused, timeout): [OabpTransportException];
 *  - **protocol** — a response arrived but could not be parsed / was malformed:
 *    [OabpSerializationException];
 *  - **API** — the server returned a non-2xx status: [OabpApiException], which exposes the
 *    status code and raw body.
 *
 * Unlike the Java SDK these are unchecked (Kotlin has no checked exceptions); suspend
 * functions document the failure modes they throw.
 */
public sealed class OabpException(
    message: String,
    cause: Throwable? = null,
) : RuntimeException(message, cause)

/** The request never produced an HTTP response (I/O error, connection refused, timeout). */
public class OabpTransportException(
    message: String,
    cause: Throwable? = null,
) : OabpException(message, cause)

/** A response arrived but could not be (de)serialized into the expected shape. */
public class OabpSerializationException(
    message: String,
    cause: Throwable? = null,
) : OabpException(message, cause)

/**
 * The OABP server returned a non-2xx HTTP status.
 *
 * @property statusCode the HTTP status code returned by the server.
 * @property body the raw response body (possibly empty, never `null`).
 */
public class OabpApiException(
    public val statusCode: Int,
    public val body: String,
    message: String,
) : OabpException(message) {
    /** `true` for 4xx statuses (the request was at fault). */
    public val isClientError: Boolean
        get() = statusCode in 400..499

    /** `true` for 5xx statuses (the server failed). */
    public val isServerError: Boolean
        get() = statusCode in 500..599

    /** `true` if the resource was not found (HTTP 404). */
    public val isNotFound: Boolean
        get() = statusCode == 404
}
